STORAGE/Refactor & Testing

boonwj opened the pull request:
This PR brings the following changes to the HoloStorageAccessor

## Refactor
- Shift configuration and blob initialisation out of the router setup function.
- Configuration will be passed in to the setup function

## Tests
- Create mock FHIR server for testing of responses 
- Add tests for all GET and PUT endpoints in HoloStorageAccessor 

Code coverage currently at ~70%. Lacking tests for blobstorage related code (e.g. Hologram DELETE / POST, blobstorage operations)

boonwj committed:
STOARGE: Moved configuration out of router

To allow setting up of testing without needing access to blob storage

boonwj committed:
STORAGE: Fix util tests to new config method

boonwj committed:
WIP: Test for api_default

boonwj committed:
STORAGE: Add GET author test in api_default

- Create test server with patient and author data
- Add test for GET author endpoint

boonwj committed:
STORAGE: Tests for multi patient & author get

boonwj committed:
STORAGE: Tests for Hologram GET endpoints

boonwj committed:
STORAGE: Test for PUT patients

- Update error message on decode error
- Add tests for Patients PUT endpoint

boonwj committed:
STORAGE: Test for PUT authors

boonwj committed:
STORAGE: Refactor Single resource GET to same func

boonwj committed:
STORAGE: Fix error with on PUT operations

- Put operations should return the API spec and not the FHIR spec
- Refactor all Put endpoints to use PutDataIntoFHIR()

boonwj committed:
STORAGE: Slight refactor and comments

nielsboecker commented on HoloStorageAccessor/internal/holostorageaccessor/api_default_test.go:
```diff
@@ -0,0 +1,1470 @@
+	"github.com/google/go-cmp/cmp"
+)
+
+// This creates a test server that mocks the FHIR server responses
```
Maybe a separate file?

boonwj replied:
shifted it to a separate file `mock_server_test.go`

nielsboecker commented on HoloStorageAccessor/internal/holostorageaccessor/api_default_test.go:
```diff
@@ -0,0 +1,1470 @@
+						}
+					],
+					"gender": "female",
+					"birthDate": "1960-01-02"
```
Damn Britney is old

boonwj replied:
absolutely

nielsboecker commented on HoloStorageAccessor/internal/holostorageaccessor/api_default_test.go:
```diff
@@ -0,0 +1,1470 @@
+			wantStatus: 400,
+		},
+		"wrong_query_field": {
+			inUrl:      "/api/v1/patients?pid=p1000",
```
Nice

nielsboecker commented on HoloStorageAccessor/internal/holostorageaccessor/api_default_test.go:
```diff
@@ -0,0 +1,1470 @@
+		},
+	}
+
+	// Start a local HTTP server and Router
```
I don't understand why you do this eight times?

boonwj replied:
Changed it such that `TestMain()` function does an initial setup before running all tests.

nielsboecker reviewed (approved):
Looks great and the test coverage now really provides some serious confidence.

The only thing I'm not sure about is that `api_default_test.go` is really long. Maybe it could be broken apart by main routes, e.g. `api_default_holograms_test.go`. The mock FHIR server could be its own file and for each test, only the relevant routes need to be mocked with return values.

We could then also consider to put all of that stuff in a `test` folder, even if it makes imports a bit harder and breaks the golang tradition of keeping it flat? To make it more concise you could even think about extracting the expected JSONs into actual `json` files.

But having said that I'm not sure if it is a good idea as I don't know the language, and I don't think you should put much more time into it necessarily.

boonwj committed:
STORAGE: TestMain() for mock server setup

Replaces the server setup call found in every test in
default_api_test.go

boonwj committed:
STORAGE: Removed unused hologram_upload model

Auto-generated by openapi generator. But the multipart form
implementation required fields to be obtained specificially. So the
form contents are mapped straight into the hologram, author and patient
models, whereas file uploads was handled separately. So there is no
longer a need for the hologram_upload model.

boonwj committed:
STORAGE: Shift mock server setup to another file

boonwj commented:
Addressed most of the concerns except for shifting the code out of the flat structure. Also found out that I have some bloated code that was removed.

nielsboecker reviewed (approved):
Double-approve
